graph/src/firehose/endpoints.rs

```diff
@@ -1,826 +1,960 @@
+    }
+}
+
+#[derive(Debug)]
+pub struct FirehoseEndpoint {
+    pub provider: ProviderName,
+    pub auth: AuthInterceptor,
+    pub filters_enabled: bool,
+    pub compression_enabled: bool,
+    pub subgraph_limit: SubgraphLimit,
+    genesis_decoder: Box<dyn GenesisDecoder>,
+    endpoint_metrics: Arc<EndpointMetrics>,
+    channel: Channel,
+}
+
+#[derive(Debug)]
+pub struct ConnectionHeaders(HashMap<MetadataKey<Ascii>, MetadataValue<Ascii>>);
+
+#[async_trait]
+impl NetIdentifiable for Arc<FirehoseEndpoint> {
+    async fn net_identifiers(&self) -> Result<ChainIdentifier, anyhow::Error> {
+        let ptr: BlockPtr = self.genesis_decoder.get_genesis_block_ptr(self).await?;
+
+        Ok(ChainIdentifier {
+            net_version: "0".to_string(),
+            genesis_block_hash: ptr.hash,
+        })
+    }
+    fn provider_name(&self) -> ProviderName {
+        self.provider.clone()
+    }
+}
+
+impl ConnectionHeaders {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+    pub fn with_deployment(mut self, deployment: DeploymentHash) -> Self {
+        if let Ok(deployment) = deployment.parse() {
+            self.0
+                .insert("x-deployment-id".parse().unwrap(), deployment);
+        }
+        self
+    }
+    pub fn add_to_request<T>(&self, request: T) -> Request<T> {
+        let mut request = Request::new(request);
+        self.0.iter().for_each(|(k, v)| {
+            request.metadata_mut().insert(k, v.clone());
+        });
+        request
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Ord, Eq, PartialOrd)]
+pub enum AvailableCapacity {
+    Unavailable,
+    Low,
+    High,
+}
+
+// TODO: Find a new home for this type.
+#[derive(Clone, Debug, PartialEq, Ord, Eq, PartialOrd)]
+pub enum SubgraphLimit {
+    Disabled,
+    Limit(usize),
+    Unlimited,
+}
+
+impl SubgraphLimit {
+    pub fn get_capacity(&self, current: usize) -> AvailableCapacity {
+        match self {
+            // Limit(0) should probably be Disabled but just in case
+            SubgraphLimit::Disabled | SubgraphLimit::Limit(0) => AvailableCapacity::Unavailable,
+            SubgraphLimit::Limit(total) => {
+                let total = *total;
+                if current >= total {
+                    return AvailableCapacity::Unavailable;
+                }
+
+                let used_percent = current * 100 / total;
+
+                // If total is low it can vary very quickly so we can consider 50% as the low threshold
+                // to make selection more reliable
+                let threshold_percent = if total <= LOW_VALUE_THRESHOLD {
+                    LOW_VALUE_USED_PERCENTAGE
+                } else {
+                    HIGH_VALUE_USED_PERCENTAGE
+                };
+
+                if used_percent < threshold_percent {
+                    return AvailableCapacity::High;
+                }
+
+                AvailableCapacity::Low
+            }
+            _ => AvailableCapacity::High,
+        }
+    }

```

Comment:
Why does this discretize the `SubgraphLimit`? I would have thought you want to select the provider with the biggest available capacity and/or the lowest usage. But it seems this message first classifies providers into 3 groups, and presumably something than picks one from one of the groups.

Reply:
This code wasn't modified by this, it was just moved, probably because of a rebase and I don't really want to open the scope of the review to that code since it's not part of the PR